packages/storefront-api-client/src/storefront-api-client.ts

```diff
@@ -0,0 +1,211 @@
+  validateRequiredAccessTokens(publicAccessToken, privateAccessToken);
+  validatePrivateAccessTokenUsage(privateAccessToken);
+
+  const apiUrlFormatter = generateApiUrlFormatter(
+    storeUrl,
+    apiVersion,
+    baseApiVersionValidationParams
+  );
+
+  const config: StorefrontApiClientConfig = {
+    storeDomain: storeUrl,
+    apiVersion,
+    publicAccessToken: publicAccessToken ?? null,
+    privateAccessToken: privateAccessToken ?? null,
+    headers: {
+      "Content-Type": DEFAULT_CONTENT_TYPE,
+      Accept: DEFAULT_CONTENT_TYPE,
+      [SDK_VARIANT_HEADER]: DEFAULT_SDK_VARIANT,
+      [SDK_VERSION_HEADER]: DEFAULT_CLIENT_VERSION,
+      ...(clientName ? { [SDK_VARIANT_SOURCE_HEADER]: clientName } : {}),
+      ...(publicAccessToken
+        ? { [PUBLIC_ACCESS_TOKEN_HEADER]: publicAccessToken }
+        : { [PRIVATE_ACCESS_TOKEN_HEADER]: privateAccessToken! }),
+    },
+    apiUrl: apiUrlFormatter(),
+    clientName,
+  };
+
+  const graphqlClient = createGraphQLClient({
+    headers: config.headers,
+    url: config.apiUrl,
+    retries,
+    fetchApi: clientFetchApi,
+    logger,
+  });
+
+  const getHeaders = generateGetHeader(config);
+  const getApiUrl = generateGetApiUrl(config, apiUrlFormatter);
+
+  const getGQLClientRequestProps = generateGetGQLClientProps({
+    getHeaders,
+    getApiUrl,
+  });
+
+  const fetch = (...props: ApiClientRequestParams) => {
+    const requestProps = getGQLClientRequestProps(...props);
+    return graphqlClient.fetch(...requestProps);
+  };
+
+  const request = <TData>(...props: ApiClientRequestParams) => {
+    const requestProps = getGQLClientRequestProps(...props);
+    return graphqlClient.request<TData>(...requestProps);
+  };
+
+  const client: StorefrontApiClient = {
+    config,
+    getHeaders,
+    getApiUrl,
+    fetch,
+    request,
+  };
+
+  return Object.freeze(client);
+}
+
+function generateApiUrlFormatter(
+  storeUrl: string,
+  defaultApiVersion: string,
+  baseApiVersionValidationParams: Omit<
+    Parameters<typeof validateApiVersion>[0],
+    "apiVersion"
+  >
+) {
+  return (apiVersion?: string) => {
+    if (apiVersion) {
+      validateApiVersion({
+        ...baseApiVersionValidationParams,
+        apiVersion,
+      });
+    }
+
+    const urlApiVersion = (apiVersion ?? defaultApiVersion).trim();
+
+    return `${storeUrl}/api/${urlApiVersion}/graphql.json`;
+  };
+}
+
+function generateGetHeader(
+  config: StorefrontApiClientConfig
+): StorefrontApiClient["getHeaders"] {
+  return (customHeaders) => {
+    return customHeaders
+      ? { ...customHeaders, ...config.headers }
+      : config.headers;
+  };
+}
+
+function generateGetApiUrl(
+  config: StorefrontApiClientConfig,
+  apiUrlFormatter: (version?: string) => string
+): StorefrontApiClient["getApiUrl"] {
+  return (propApiVersion?: string) => {
+    return propApiVersion ? apiUrlFormatter(propApiVersion) : config.apiUrl;
+  };
+}
+
+function generateGetGQLClientProps({
```

Comment:
This might be universal enough to move into the base client's utils...will look more into this next week.